packages/snaps-controllers/src/cronjob/CronjobController.ts

```diff
@@ -195,12 +215,70 @@ export class CronjobController extends BaseController<
       `${controllerName}:getBackgroundEvents`,
       (...args) => this.getBackgroundEvents(...args),
     );
+  }
 
-    this.dailyCheckIn().catch((error) => {
-      logError(error);
+  /**
+   * Execute a request.
+   *
+   * @param snapId - ID of a Snap.
+   * @param request - Request to be executed.
+   */
+  async #executeRequest(snapId: SnapId, request: unknown) {
+    await this.messagingSystem
+      .call('SnapController:handleRequest', {
+        snapId,
+        origin: '',
+        handler: HandlerType.OnCronjob,
+        request: request as Record<string, unknown>,
+      })
+      .catch((error) => {
+        logError(error);
+      });
+  }
+
+  /**
+   * Setup a timer.
+   *
+   * @param id - ID of a timer.
+   * @param ms - Time in milliseconds.
+   * @param snapId - ID of a Snap.
+   * @param onComplete - Callback function to be executed when the timer completes.
+   * @param validateTime - Whether the time should be validated.
+   */
+  #setupTimer(
+    id: string,
+    ms: number,
+    snapId: SnapId,
+    onComplete: () => void,
+    validateTime = true,
+  ) {
+    if (validateTime && ms <= 0) {
+      throw new Error('Cannot schedule execution in the past.');
+    }
+
+    const timer = new Timer(ms);
+    timer.start(() => {
+      onComplete();
+      this.#timers.delete(id);
+      this.#snapIds.delete(id);
     });
 
-    this.#rescheduleBackgroundEvents(Object.values(this.state.events));
+    this.#timers.set(id, timer);
+    this.#snapIds.set(id, snapId);
+  }
+
+  /**
+   * Cleanup a timer.
+   *
+   * @param id - ID of a timer.
+   */
+  #cleanupTimer(id: string) {
+    const timer = this.#timers.get(id);
+    if (timer) {
```

Comment:
Should this throw an error if the timer doesn't exist?

Reply:
Hmm I don't think so, this function is only called when cancelling a background event.

Reply:
https://github.com/MetaMask/snaps/pull/3029/files#diff-d3d7c575b1031aeae16afee89d5c29a141c7edc55f190c5c0bc8d0071e9fcf2fR345-R353